### DISCUSSION TITLE
Can Python code be called from DuckDB SQL queries?

### DISCUSSION THREAD

**Discussion by parvizfarnia** [duckdb/duckdb#6454] [Q&A] [answered]
Hi,

I wasn't sure to create this thread as "`Q&A`" or "`Ideas`" as I don't know whether the functionality already exists in `DuckDB`.

Using `PostgreSQL` there is a very useful extension called `plpython3u`; that allows to call user defined Python function in SQL queries.  

Here is the link : [PL/Python Functions](https://www.postgresql.org/docs/current/plpython-funcs.html)

I would like to know whether the same can be done in DuckDB?

Thanks.

**Comment by l1t1** [comment]
how about `import duckdb` in python?

**Reply by parvizfarnia** [reply]
I already use DuckDB in my python programs via `import duckdb` but what I asked, was to see the possibilities to directly incorporate Python functions at SQL level as indicated by the link I provided above.

**Comment by l1t1** [comment]
it can be call in following code
```
con = duckdb.connect(database=':memory:')

con.execute("CREATE TABLE items(item VARCHAR, value DECIMAL(10,2), count INTEGER)")

con.execute("INSERT INTO items VALUES ('jeans', 20.0, 1), ('hammer', 42.2, 2)")

con.execute("SELECT * FROM items where upper(item)=?",['jeans'.upper()]);

print(con.fetchall())
```

**Comment by Alex-Monahan** [comment, marked as answer]
They can't be called from SQL just yet, but DuckDB's relational API supports table functions in Python!
https://github.com/duckdb/duckdb/discussions/4797
